text/0694-validate-args.md

```diff
@@ -0,0 +1,422 @@
+with the `prop-types` like APIs. However, there are still cases which TypeScript
+cannot solve:
+
+1. Providing API safety/guarantees for plain JavaScript users. This is true even
+   for TypeScript addon authors, as they may have plain JavaScript users.
+2. Validating non-type invariants, such as args that are relative to each other
+   (e.g. `@max` must be larger than `@min`)
+
+So even for TypeScript users, argument validation can still provide benefits.
+Between this, and the fact that Ember is committed to providing a first-class
+user experience to JavaScript users, it still makes sense to provide this hook
+independent of any potential TypeScript support.
+
+## Terminology
+
+For the remainder of this RFC, the term "invokable" will be used to refer to
+component/modifier/helper, since all three will be validated in the same way
+without any differences.
+
+## Detailed design
+
+The `validateArgs` hook will be added to each of the invokable manager
+interfaces. It will have the following signature in each case:
+
+```ts
+interface Arguments {
+  named: Record<string, unknown>;
+  positional: unknown[];
+}
+
+interface InvokableManager {
+  validateArgs(definition: object, args: Arguments): void;
+}
+```
+
+This hook receives the definition of the invokable as its first
+argument, and the args passed as the second argument. It should not return any
+values, and should throw an error if the arguments are not valid. It also will
+only run in development builds, and will be completely inert otherwise.
+
+This API design accomplishes two main things:
+
+1. It requires users to associate their validations with the _definition_ of the
+   invokable, rather than the instance of it. This will prevent users from
+   developing APIs that rely on instance state, and allows validations to be
+   associated with invokables which do not have instance state, such as
+   Template-only components. If it is determined that instance state would be
+   useful in validations, then it can be added in a future RFC via a new
+   capabilities version.
+2. It enables general validations against the shape of arguments, rather than
+   specific argument values. For instance, users can assert that extra arguments
+   are not passed to a component, only the required ones. They can also assert
+   relative values, such as `@max` being greater than `@min`.
+
+This hook will be called once when the component is first invoked, and is
+autotracked. It will be called again whenever the values that are validated and
+tracked change, _on the next argument access_. This means that if the arguments
+are updated, but no arguments are used at all within the component, then
+validations _may not_ run again. This is due to the fact that argument usage is
+lazy, and so we cannot know the values until they are used.
+
+While `validateArgs` will be autotracked on its own, it will not be _consumed_
+by the external tracking contexts. This prevents it from impacting the tracking
```

Comment:
We have the ability to disable autotracking contexts internally, and use it for various reasons (such as backwards compatibility). We have not exposed this capability, and generally don't think we will do so as in most cases it doesn't make sense, but I think here since this is a DEBUG only feature it may make sense. It does mean that validations may not rerun when you think they would, but it also prevents behavior from diverging too much in production.